File: cpp/samples/transport/shipid.cpp
/**
 * @file shipid.cpp
 * @brief Demonstrates usage of vista-sdk ShipId API
 * @details This sample shows how to construct, parse, and inspect ShipId values,
 *          which identify a vessel in ISO 19848 transport packages either by IMO number
 *          or by an alternative identifier.
 */

#include <dnv/vista/sdk/core/ImoNumber.h>
#include <dnv/vista/sdk/transport/ShipId.h>

#include <iostream>
#include <string>
#include <vector>

int main()
{
    using namespace dnv::vista::sdk;
    using namespace dnv::vista::sdk::transport;

    std::cout << "=== vista-sdk ShipId Sample ===\n\n";

    {
        std::cout << "1. ShipId: Constructing from an IMO number\n";
        std::cout << "----------------------------------------------\n";

        auto imo = ImoNumber::fromString("9074729");
        if (imo.has_value())
        {
            ShipId shipId{ *imo };

            std::cout << "  isImoNumber: " << std::boolalpha << shipId.isImoNumber() << "\n";
            std::cout << "  isOtherId  : " << shipId.isOtherId() << "\n";
            std::cout << "  toString() : " << shipId.toString() << "\n";
        }
        std::cout << "\n";
    }

    {
        std::cout << "2. ShipId: Constructing from an alternative identifier\n";
        std::cout << "-----------------------------------------------------------\n";

        ShipId shipId{ std::string{ "VESSEL-XYZ-789" } };

        std::cout << "  isImoNumber: " << std::boolalpha << shipId.isImoNumber() << "\n";
        std::cout << "  isOtherId  : " << shipId.isOtherId() << "\n";
        std::cout << "  toString() : " << shipId.toString() << "\n";
        std::cout << "\n";
    }

    {
        std::cout << "3. ShipId::fromString: Parsing header ShipID values\n";
        std::cout << "--------------------------------------------------------\n";

        std::vector<std::string> testValues = {
            "IMO9074729",     // valid IMO number, with prefix
            "imo9074729",     // valid IMO number, case-insensitive prefix
            "IMO1234568",     // "IMO" prefix but failing checksum -> falls back to alternative id
            "VESSEL-ABC-123", // plain alternative identifier
            ""                // empty -> no ShipId
        };

        for (const auto& value : testValues)
        {
            auto shipId = ShipId::fromString(value);

            std::cout << "  '" << value << "'\n";
            if (!shipId.has_value())
            {
                std::cout << "    -> (invalid, empty input)\n";
                continue;
            }

            if (shipId->isImoNumber())
            {
                std::cout << "    -> IMO number: " << shipId->toString() << "\n";
            }
            else
            {
                std::cout << "    -> Alternative id: " << shipId->toString() << "\n";
            }
        }
        std::cout << "\n";
    }

    {
        std::cout << "4. ShipId::match: Pattern matching on the identifier kind\n";
        std::cout << "----------------------------------------------------------------\n";

        std::vector<ShipId> shipIds;
        if (auto imo = ImoNumber::fromString("9785811"); imo.has_value())
        {
            shipIds.emplace_back(*imo);
        }
        shipIds.emplace_back(std::string{ "BARGE-042" });

        for (const auto& shipId : shipIds)
        {
            auto description = shipId.match(
                [](const ImoNumber& imoNumber) { return "IMO number  : " + imoNumber.toString(); },
                [](std::string_view otherId) { return "Alternative : " + std::string{ otherId }; });

            std::cout << "  " << description << "\n";
        }
        std::cout << "\n";
    }

    {
        std::cout << "5. ShipId: Equality comparison\n";
        std::cout << "----------------------------------\n";

        auto imo1 = ImoNumber::fromString("9074729");
        auto imo2 = ImoNumber::fromString("9074729");
        auto imo3 = ImoNumber::fromString("9785811");

        if (!imo1.has_value() || !imo2.has_value() || !imo3.has_value())
        {
            return 1;
        }

        ShipId shipIdA{ *imo1 };
        ShipId shipIdB{ *imo2 };
        ShipId shipIdC{ *imo3 };

        std::cout << std::boolalpha;
        std::cout << "  IMO9074729 == IMO9074729 : " << (shipIdA == shipIdB) << "\n";
        std::cout << "  IMO9074729 == IMO9785811 : " << (shipIdA == shipIdC) << "\n";
        std::cout << "\n";
    }

    return 0;
}

File: bindings/rust/vista-sdk/examples/transport/ship_id.rs
//! Demonstrates usage of the vista-sdk ShipId API.
//!
//! Shows how to construct, parse, and inspect ShipId values, which identify a vessel
//! in ISO 19848 transport packages either by IMO number or by an alternative identifier.

use vista_sdk::core::imo_number::ImoNumber;
use vista_sdk::transport::ship_id::ShipId;

fn main() {
    println!("=== vista-sdk ShipId Sample ===\n");

    {
        println!("1. ShipId: Constructing from an IMO number");
        println!("----------------------------------------------");

        if let Some(imo) = ImoNumber::from_str("9074729") {
            let ship_id = ShipId::from_imo_number(imo);

            println!("  is_imo_number: {}", ship_id.is_imo_number());
            println!("  is_other_id  : {}", ship_id.is_other_id());
            println!("  to_string()  : {}", ship_id);
        }
        println!();
    }

    {
        println!("2. ShipId: Constructing from an alternative identifier");
        println!("-----------------------------------------------------------");

        let ship_id = ShipId::from_other_id("VESSEL-XYZ-789").expect("non-empty string");

        println!("  is_imo_number: {}", ship_id.is_imo_number());
        println!("  is_other_id  : {}", ship_id.is_other_id());
        println!("  to_string()  : {}", ship_id);
        println!();
    }

    {
        println!("3. ShipId::from_string: Parsing header ShipID values");
        println!("--------------------------------------------------------");

        let test_values = [
            "IMO9074729",     // valid IMO number, with prefix
            "imo9074729",     // valid IMO number, case-insensitive prefix
            "IMO1234568",     // "IMO" prefix but failing checksum -> falls back to alternative id
            "VESSEL-ABC-123", // plain alternative identifier
            "",               // empty -> None
        ];

        for value in &test_values {
            print!("  '{value}'\n");
            match ShipId::from_string(value) {
                None => println!("    -> (invalid, empty input)"),
                Some(ship_id) if ship_id.is_imo_number() => {
                    println!("    -> IMO number: {ship_id}");
                }
                Some(ship_id) => {
                    println!("    -> Alternative id: {ship_id}");
                }
            }
        }
        println!();
    }

    {
        println!("4. ShipId: Pattern matching on the identifier kind");
        println!("------------------------------------------------------");

        let mut ship_ids: Vec<ShipId> = Vec::new();
        if let Some(imo) = ImoNumber::from_str("9785811") {
            ship_ids.push(ShipId::from_imo_number(imo));
        }
        ship_ids.push(ShipId::from_other_id("BARGE-042").expect("non-empty string"));

        for ship_id in &ship_ids {
            let description = match ship_id {
                ShipId::Imo(imo) => format!("IMO number  : {imo}"),
                ShipId::Other(id) => format!("Alternative : {id}"),
            };
            println!("  {description}");
        }
        println!();
    }

    {
        println!("5. ShipId: Equality comparison");
        println!("----------------------------------");

        let imo1 = ImoNumber::from_str("9074729").expect("9074729 is valid");
        let imo2 = ImoNumber::from_str("9074729").expect("9074729 is valid");
        let imo3 = ImoNumber::from_str("9785811").expect("9785811 is valid");

        let ship_id_a = ShipId::from_imo_number(imo1);
        let ship_id_b = ShipId::from_imo_number(imo2);
        let ship_id_c = ShipId::from_imo_number(imo3);

        println!("  IMO9074729 == IMO9074729 : {}", ship_id_a == ship_id_b);
        println!("  IMO9074729 == IMO9785811 : {}", ship_id_a == ship_id_c);
        println!();
    }
}
